connection.go

```diff
@@ -0,0 +1,239 @@
+		handlers:   handlers,
+		namespaces: newNamespaces(),
+	}
+
+	if err := ret.connect(); err != nil {
+		_ = ret.Close()
+		return err
+	}
+
+	return nil
+}
+
+func (c *conn) Close() error {
+	var err error
+
+	c.closeOnce.Do(func() {
+		// for each namespace, leave all rooms, and call the disconnect handler.
+		c.namespaces.Range(func(ns string, nc *namespaceConn) {
+			nc.LeaveAll()
+
+			if nh, _ := c.handlers.Get(ns); nh != nil && nh.onDisconnect != nil {
+				nh.onDisconnect(nc, clientDisconnectMsg)
+			}
+		})
+		err = c.Conn.Close()
+
+		close(c.quitChan)
+	})
+
+	return err
+}
+
+func (c *conn) connect() error {
+	rootHandler, ok := c.handlers.Get(rootNamespace)
+	if !ok {
+		return errUnavailableRootHandler
+	}
+
+	root := newNamespaceConn(c, aliasRootNamespace, rootHandler.broadcast)
+	c.namespaces.Set(rootNamespace, root)
+
+	root.Join(root.ID())
+
+	c.namespaces.Range(func(ns string, nc *namespaceConn) {
+		nc.SetContext(c.Conn.Context())
+	})
+
+	header := parser.Header{
+		Type: parser.Connect,
+	}
+
+	if err := c.encoder.Encode(header); err != nil {
+		return err
+	}
+
+	handler, ok := c.handlers.Get(header.Namespace)
+
+	go c.serveError()
+	go c.serveWrite()
+	go c.serveRead()
+
+	if ok {
+		_, err := handler.dispatch(root, header)
+		return err
+	}
+
+	return nil
+}
+
+func (c *conn) nextID() uint64 {
+	c.id++
+
+	return c.id
+}
+
+func (c *conn) write(header parser.Header, args ...reflect.Value) {
+	data := make([]interface{}, len(args))
+
+	for i := range data {
+		data[i] = args[i].Interface()
+	}
+
+	pkg := parser.Payload{
+		Header: header,
+		Data:   data,
+	}
+
+	select {
+	case c.writeChan <- pkg:
+	case <-c.quitChan:
+		return
+	}
+}
+
+func (c *conn) onError(namespace string, err error) {
+	select {
+	case c.errorChan <- newErrorMessage(namespace, err):
+	case <-c.quitChan:
+		return
+	}
+}
+
+func (c *conn) serveError() {
+	defer c.Close()
+
+	for {
+		select {
+		case <-c.quitChan:
+			return
+		case err := <-c.errorChan:
+			errMsg, ok := err.(errorMessage)
+			if !ok {
+				continue
+			}
+
+			if handler := c.namespace(errMsg.namespace); handler != nil {
+				if handler.onError != nil {
+					nsConn, ok := c.namespaces.Get(errMsg.namespace)
+					if !ok {
+						continue
+					}
+					handler.onError(nsConn, errMsg.err)
+				}
+			}
+		}
+	}
+}
+
+func (c *conn) serveWrite() {
+	defer c.Close()
+
+	for {
+		select {
+		case <-c.quitChan:
+			return
+		case pkg := <-c.writeChan:
+			if err := c.encoder.Encode(pkg.Header, pkg.Data); err != nil {
+				c.onError(pkg.Header.Namespace, err)
+			}
+		}
+	}
+}
+
+func (c *conn) serveRead() {
+	defer c.Close()
+
+	var event string
+
+	for {
+		var header parser.Header
+
+		if err := c.decoder.DecodeHeader(&header, &event); err != nil {
+			c.onError(rootNamespace, err)
+			return
+		}
+
+		if header.Namespace == aliasRootNamespace {
+			header.Namespace = rootNamespace
+		}
+
+		var err error
+		switch header.Type{
+		case parser.Ack, parser.Connect, parser.Disconnect:
```

Comment:
why we not handle `parser.Error`?

Reply:
You are right, it should be here too.

Reply:
Issue https://github.com/googollee/go-socket.io/issues/466